server/server.go

```diff
@@ -0,0 +1,393 @@
+func addFile(ctx context.Context, peer *ipfslite.Peer, addParams *pb.AddParams, reader io.Reader, ch chan addFileResult) {
+	defer close(ch)
+	params := &ipfslite.AddParams{
+		Layout:    addParams.GetLayout(),
+		Chunker:   addParams.GetChunker(),
+		RawLeaves: addParams.GetRawLeaves(),
+		Hidden:    addParams.GetHidden(),
+		Shard:     addParams.GetShared(),
+		NoCopy:    addParams.GetNoCopy(),
+		HashFun:   addParams.GetHashFun(),
+	}
+	node, err := peer.AddFile(ctx, reader, params)
+	if err != nil {
+		ch <- addFileResult{Err: err}
+		return
+	}
+	ch <- addFileResult{Node: node}
+}
+
+func (s *ipfsLiteServer) AddFile(srv pb.IpfsLite_AddFileServer) error {
+	req, err := srv.Recv()
+	if err != nil {
+		return err
+	}
+	var addParams *pb.AddParams
+	switch payload := req.GetPayload().(type) {
+	case *pb.AddFileRequest_AddParams:
+		addParams = payload.AddParams
+	default:
+		return fmt.Errorf("expexted AddParams for AddFileRequest.Payload but got %T", payload)
+	}
+
+	reader, writer := io.Pipe()
+
+	addFileChannel := make(chan addFileResult)
+	go addFile(srv.Context(), s.peer, addParams, reader, addFileChannel)
+
+	for {
+		req, err := srv.Recv()
+		if err == io.EOF {
+			writer.Close()
+			break
+		} else if err != nil {
+			writer.CloseWithError(err)
+			break
+		}
+		switch payload := req.GetPayload().(type) {
+		case *pb.AddFileRequest_Chunk:
+			_, writeErr := writer.Write(payload.Chunk)
+			if writeErr != nil {
+				return writeErr
+			}
+		default:
+			return fmt.Errorf("expected Chunk for AddFileRequest.Payload but got %T", payload)
+		}
+	}
+
+	addFileResult := <-addFileChannel
+	if addFileResult.Err != nil {
+		return addFileResult.Err
+	}
+
+	respNode, err := nodeToPbNode(addFileResult.Node)
+	if err != nil {
+		return err
+	}
+
+	return srv.SendAndClose(&pb.AddFileResponse{Node: respNode})
+}
+
+func (s *ipfsLiteServer) GetFile(req *pb.GetFileRequest, srv pb.IpfsLite_GetFileServer) error {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return err
+	}
+
+	reader, err := s.peer.GetFile(srv.Context(), cid)
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, getFileChunkSize)
+	for {
+		size, err := reader.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		srv.Send(&pb.GetFileResponse{Chunk: buffer[:size]})
+		if err == io.EOF {
+			return nil
+		}
+	}
+}
+
+func (s *ipfsLiteServer) HasBlock(ctx context.Context, req *pb.HasBlockRequest) (*pb.HasBlockResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	hasBlock, err := s.peer.HasBlock(cid)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.HasBlockResponse{HasBlock: hasBlock}, nil
+}
+
+func (s *ipfsLiteServer) AddNode(ctx context.Context, req *pb.AddNodeRequest) (*pb.AddNodeResponse, error) {
+	cid, err := cid.Decode(req.Block.GetCid())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: there is also just blocks.NewBlock()
+	block, err := blocks.NewBlockWithCid(req.Block.GetRawData(), cid)
```

Comment:
At this point, I have the client sending a cid string and the raw node bytes. I'm using both of those bits of info to reconstruct the Block and Node here on the server. I'm a bit confused because you can also do this without the cid (I'm not sure how it knows how the bytes are encoded in that case). Anyway, open to suggestions about the right way to do this.

Reply:
The raw bytes from the client can be encoded using cbor or Protobufs or any other supported format I guess.

Reply:
This also raises questions about how clients in different languages will create the raw bytes to send. There looks to be a ipld cbor lib for Java but not Obj C, for example.

Reply:
I believe the cid comes from the raw bytes, which use multicodec to inform readers how they are encoded: https://github.com/multiformats/multicodec

Reply:
I'm always forgetting which multi thing does what. This helps: https://github.com/multiformats/multiformats/blob/master/img/multiformats.001.jpg

Reply:
Looking at this again, it's probably equally common to add a block without the cid. In that case, it's just generated automatically. (without CID -> you added some new data, with cid -> you're adding a block you received by requesting it from the network using its cid, as in, you already have it)

Reply:
So, should we just treat the cid param as optional and call the corresponding method depending on if it is set?

Reply:
SGTM